### ISSUE TITLE
Inducing point behavior

### ISSUE THREAD

**Issue by mbronckers** [mbronckers/GI-PVI#3]
**Question**: are the inducing points `zs` and associated pseudo outputs `yz` optimized correctly?

This issue serves as a log to determine whether the behavior of the GI-BNN is as expected under various scenarios

# Baseline Scenario
**Scenario**: z initialized to first batch of training points, 1000 epochs, 1000 training points, batch=100, inducing M = 100

**Results**:
- The Z move slightly away from the initialization, which is probably due to the KL prior term.

<img width="1231" alt="image" src="https://user-images.githubusercontent.com/25270093/171630621-e662de88-8ade-430f-8455-c4484ce744c4.png">

**Comment by mbronckers** [comment]
**Scenario**: random z initialization on N(0, 1), 2000 epochs, 1000 training points, lr 1e-3

**Results**: 
- Lower KL => expected since KL(q,p) is wrt p = N(0,1)
- Runtime: 1m

**Plots**
<img width="1693" alt="image" src="https://user-images.githubusercontent.com/25270093/171588307-8550e8f1-6419-463a-b423-7ac618c6b260.png">
<img width="500" alt="image" src="https://user-images.githubusercontent.com/25270093/171588209-4ae5b51b-343c-4d4f-bcfe-de1388c230aa.png">

**Comment by mbronckers** [comment]
I have written boilerplate code to save, plot, and log the experiment in the results/ folder when running locally. You can force push results to github, as done here: [example experiment](https://github.com/mbronckers/GI-PVI/tree/main/results/2022-06-07-08.47.45_eval-all).

**Comment by mbronckers** [comment]
**Scenario**:  no prior KL term in ELBO estimate. Otherwise standard configuration and seed = 1.

**Results**:
- The inducing points updates away from training data are still happening, so this is not due to the prior KL term.

**Plots**:
<img width="500" alt="image" src="https://user-images.githubusercontent.com/25270093/172344504-460cc946-46ed-449b-9b8f-b82ee5c72b4c.png">

<img width="500" alt="image" src="https://user-images.githubusercontent.com/25270093/172344599-a2e1d108-d322-4593-8caf-65f27d9ecf78.png">

**Comment by mbronckers** [comment]
**Scenario**: random inducing point initialization. Otherwise standard configuration and seed = 1. [Experiment results](https://github.com/mbronckers/GI-PVI/tree/main/results/2022-06-07-09.21.00_rand-z)

**Plots**:

<img width="655" alt="image" src="https://user-images.githubusercontent.com/25270093/172346001-071e7c4d-9d25-4474-a0f8-da2581714827.png">
<img width="619" alt="image" src="https://user-images.githubusercontent.com/25270093/172346035-6f25cf31-5fb6-403a-8fb8-4d37d448d4db.png">
<img width="653" alt="image" src="https://user-images.githubusercontent.com/25270093/172346083-0662a81f-fcc5-47ff-ae1a-fd19713fb502.png">

**Comment by mbronckers** [comment]
**Scenario**: using 2nd regression dgp. 5000 epochs, random z init, 100 inducing pionts, seed = 1

<img width="1732" alt="image" src="https://user-images.githubusercontent.com/25270093/172373928-f7394069-1cc9-439e-9604-a9747b1c9ce4.png">
<img width="851" alt="image" src="https://user-images.githubusercontent.com/25270093/172373951-a877383f-79f5-42a9-8017-14e3e39030cb.png">

**Comment by mbronckers** [comment]
I cannot seem to figure out why the mean predictions are practically straight lines, across all experiments; there’s no curvature to the mean prediction across the domain.

**Comment by mbronckers** [comment]
> I cannot seem to figure out why the mean predictions are practically straight lines, across all experiments; there’s no curvature to the mean prediction across the domain.
- This was because of the absence of bias vectors in the BNN, which are now implemented:

<img width="578" alt="image" src="https://user-images.githubusercontent.com/25270093/172816039-acfae43e-5e49-45f6-9d89-9e1110bcb453.png">

The bias vectors added curvature to the mean predictions across the domain.

Implementing the NealPrior fixed the inducing point behavior.

**Comment by mbronckers** [comment]
<img width="500" alt="image" src="https://user-images.githubusercontent.com/25270093/172730772-5240aafe-2113-4ed1-a909-78099ca3489c.png">

**Comment by mbronckers** [comment]
Reopening this as the fit on DGP_2 is not great. 

The inducing points still move away from the training data:

<img width="490" alt="image" src="https://user-images.githubusercontent.com/25270093/172820791-bb5574a5-7eb9-4b9d-a8a0-367c0344a8f6.png">
